src/hb-ot-cpal-table.hh

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright © 2016  Google, Inc.
+ * Copyright © 2018  Ebrahim Byagowi
+ *
+ *  This is part of HarfBuzz, a text shaping library.
+ *
+ * Permission is hereby granted, without written agreement and without
+ * license or royalty fees, to use, copy, modify, and distribute this
+ * software and its documentation for any purpose, provided that the
+ * above copyright notice and the following two paragraphs appear in
+ * all copies of this software.
+ *
+ * IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE TO ANY PARTY FOR
+ * DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
+ * ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN
+ * IF THE COPYRIGHT HOLDER HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * THE COPYRIGHT HOLDER SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING,
+ * BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ *
+ * Google Author(s): Sascha Brawer
+ */
+
+#ifndef HB_OT_CPAL_TABLE_HH
+#define HB_OT_CPAL_TABLE_HH
+
+#include "hb-open-type-private.hh"
+
+
+namespace OT {
+
+/*
+ * Color Palette
+ * http://www.microsoft.com/typography/otspec/cpal.htm
+ */
+
+#define HB_OT_TAG_CPAL HB_TAG('C','P','A','L')
+
+struct ColorRecord
+{
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    // We do not enforce alpha != 0 because zero alpha is bogus but harmless.
+    TRACE_SANITIZE (this);
+    return_trace (true);
+  }
+
+  public:
+  HBUINT8 blue;
+  HBUINT8 green;
+  HBUINT8 red;
+  HBUINT8 alpha;
+
+  DEFINE_SIZE_STATIC (4);
+};
+
+struct CPALV1Tail
+{
+  friend struct CPAL;
+
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    return_trace (paletteFlags.sanitize (c, this) &&
+    		  paletteLabel.sanitize (c, this) &&
+    		  paletteEntryLabel.sanitize (c, this));
+  }
+
+  private:
+  inline hb_ot_color_palette_flags_t get_palette_flags (const void *base, unsigned int palette) const
+  {
+    if (paletteFlags.is_null ())
+      return HB_OT_COLOR_PALETTE_FLAG_DEFAULT;
+
+    const HBUINT32* flags = (const HBUINT32 *) (const void*) &paletteFlags (base);
```

Comment:
I wish I could use UnsizedArrayOf...

Reply:
Yeah, will move UnsizedArrayOf to hb-open-type.hh some time.